### ISSUE TITLE
dnn_face_recognition_ex face descriptor normalization

### ISSUE THREAD

**Issue by justadudewhohacks** [davisking/dlib#1304]
Hi,

I have a question about the dnn architecture employed in the [dnn_face_recognition_ex.cpp](https://github.com/davisking/dlib/blob/master/examples/dnn_face_recognition_ex.cpp) example. It looks like the output descriptors are normalized, as the they contain values between -1 and 1. 

However, I cannot see any kind of normalization applied, as the last layer is a fully connected layer. Is softmax or something applied under the hood? Because when I export the net to a caffe model, with the tool you provided, the last layer is a simple InnerProduct layer, which doesn't seem to do any kind of normalization on the final vector:
`n.fc_no_bias1 = L.InnerProduct(n.avg_pool2, num_output=128, bias_term=False);`

Sorry if this is kind of a noob question, but I am new to all the deep learning, caffe and tensorflow mess and I am trying to understand the architecture of your net.

The reason I am asking is, I am trying to port the architecture to tensorflow.js, in order to run face-recognition in the browser via WebGL, but I am trying to figure out how to build the output layer.

Thanks in advance for your help, I would highly appreciate your advice.

**Comment by davisking** [comment]
There aren't any hidden layer or any kind of normalization.  The network
functions just as it seems and you appear to have the right idea about it's
execution.  The output numbers are just by chance in that range.

**Comment by davisking** [comment]
I don't think there is any mistake in the warning message.  It just is what
it is.  Caffe isn't going to be identical to dlib in the way that all
software is not identical.

There is a deeper question about why you don't just call dlib?  Then you
wouldn't need to do all this stuff.

**Comment by justadudewhohacks** [comment]
Okay, actually my question was, whether in the c++ implementation same or valid convolution is applied, e.g. what the correct sizes of the blobs are, as this might make a difference for the output descriptor (or maybe it's not that significant and my question is just dumb, as I said I am a newbie in that area). I guess I have to dig deeper into the c++ code and eventually compare the output of individual layers, to find out where the issue resides.

Don't get me wrong, I highly appreciate your library, thanks for providing your models to the open source community! The face recognition solution provided in dlib is the best open source solution I found so far that's why I am using it for quite some time now. Even wrote an [npm package](https://github.com/justadudewhohacks/face-recognition.js), that wraps the dlib face detection, landmark and recognition API  for node.js.

But now I also wanted to try to port this to the browser, as inference with WebGL (via tensorflow.js for example, which has been released recently) seems to be insanely fast.

**Comment by davisking** [comment]
I don't think there is an issue.  The sizes are very likely what they say
they are. Maybe you found a bug and the sizes aren't what is being
reported, but that is unlikely.

Oh right, I didn't reread the thread and forgot you were the guy trying to
run it in javascript :)

**Comment by justadudewhohacks** [comment]
Alright. Is it possible by any chance to retreive the output blob for intermediate layers or only for the output layer?

**Comment by davisking** [comment]
Yes, you can get the output by just going to the layer and calling get_output().  There is a discussion of this in the introductory example programs.

**Comment by justadudewhohacks** [comment]
Thanks, this will help me a lot!

Just one last question :D. I am comparing the outputs of indivdiual layers now by writing the data of the tensor obtained from get_output() to a file as follows:

``` c++
std::vector<float> out;
for (auto iter = dlib::layer<130>(net).get_output().begin(); iter != dlib::layer<130>(net).get_output().end(); ++iter)
  out.push_back(*iter);

ofstream file("conv32_in.dat", ios::binary);
file.write((char*)out.data(), sizeof(float) * out.size());
file.close();
```

The thing I am noticing however is, that the ouput data from `dlib::layer<130>`(first conv) and `dlib::layer<128>`(first conv -> affine -> relu) is the same. Are they pointing to the same tensor or am I doing something wrong here?

**Comment by davisking** [comment]
Some layers are in-place, meaning they operate on the tensor that comes
into them rather than allocating a new one.  Any layer that can operate
in-place does, like relu and affine.

**Comment by justadudewhohacks** [comment]
I see. Awesome, thanks for your help!
